javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll

```diff
@@ -69,7 +69,26 @@ abstract class AtmConfig extends string {
    * Holds if the candidate sink `candidateSink` predicted by the machine learning model should be
    * an effective sink, i.e. one considered as a possible sink of flow in the boosted query.
    */
-  predicate isEffectiveSink(JS::DataFlow::Node candidateSink) { none() }
+  final predicate isEffectiveSink(JS::DataFlow::Node candidateSink) {
+    not exists(getAReasonSinkExcluded(candidateSink))
+  }
+
+  final EndpointCharacteristics::EndpointCharacteristic getAReasonSinkExcluded(
+    JS::DataFlow::Node candidateSink
+  ) {
+    // An endpoint is an effective sink if it has neither standard endpoint filter characteristics nor endpoint filter
+    // characteristics that are specific to this sink type.
+    exists(EndpointCharacteristics::StandardEndpointFilterCharacteristic standardFilter |
+      standardFilter.getEndpoints(candidateSink) and
+      result = standardFilter
+    )
+    or
+    exists(EndpointCharacteristics::EndpointFilterCharacteristic specificFilter |
+      specificFilter.getEndpoints(candidateSink) and
+      specificFilter.getImplications(getASinkEndpointType(), false, _) and
+      result = specificFilter
+    )
```

Comment:
This pattern of having to downcast `result` to two possible subtypes of `EndpointCharacteristic` is a little unfortunate (in any object-oriented language), and suggests we should be making more use of the ability to define predicates within the object-oriented type hierarchy.

Ideally you would just say something like:
```ql
result.isEndpointForType(candidateSink, this.getASinkEndpointType())
```
and define/override that predicate correctly on both `EndpointFilterCharacteristic` and `StandardEndpointFilterCharacteristic`.

This change would be a bit invasive, so let's come back to this one separately.

Reply:
This is just an artifact of wanting to perfectly reproduce the existing data for now, while building a class structure that makes sense for the long term. In the long term I envision `getAReasonSinkExcluded` simply saying that we exclude endpoints whose class is known with medium-high confidence a priory (ie, we only score Unknown endpoints). That can be written just by querying the implications across all EndpointCharacteristics. See the second checklist item [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2126#issue-1450274932).

Reply:
Sounds good.